### PR TITLE
miscellaneous elbo/test cleanup

### LINKED ISSUE #816: tracegraph_elbo related cleanup
- [x] refactor vectorized_map_data_info and related data structures
- [x] refactor way prune_subsample_sites etc. is used to setup a trace for computation
- [x] rethink any needed asserts/warnings in tracegraph_elbo or trace_poutine
- [x] go through test_valid_models and make sure we have good coverage
- [x] rethink integration/gradient test coverage (where possibly prefer cheaper tests)

### PR DESCRIPTION
**Pull request by martinjankowiak**
- fattened docstrings for trace_elbo and tracegraph_elbo
- removed tests/distributions/test_transformed_distribution.py
   this shouldn't be needed anymore. please confirm @neerajprad @fritzo 
- removed lognormal tests in test_inference.py and test_tracegraph_elbo.py whose main 
   purpose in life was to test transformeddistribution
- reduced number of @pytest.mark.parametrize args in test_elbo_mapdata.py

closes #816

### CODE REVIEW

File: pyro/infer/trace_elbo.py [hunk 1]
```diff
@@ -26,7 +26,22 @@ def _compute_log_r(model_trace, guide_trace):
...
     """
-    A trace implementation of ELBO-based SVI
+    A trace implementation of ELBO-based SVI. The estimator is constructed
+    along the lines of references [1] and [2]. There are no restrictions on the
+    dependency structure of the model or the guide. The gradient estimator includes
+    partial Rao-Blackwellization for reducing the variance of the estimator when
+    non-reparameterizable random variables are present. The Rao-Blackwellization is
+    partial in that it only uses conditional independence information that is marked
+    by :class:`~pyro.iarange` contexts. For more fine-grained Rao-Blackwellization,
+    see :class:`TraceGraph_ELBO`.
```
Comment (fritzo) [inline pyro/infer/trace_elbo.py]: Thanks for adding nice links! I think you'll need to be explicit about the module:
```
:class:`~pyro.infer.tracegraph_elbo.TraceGraph_ELBO`
```

File: pyro/infer/tracegraph_elbo.py [hunk 2]
```diff
@@ -170,9 +170,14 @@ class TraceGraph_ELBO(ELBO):
     A TraceGraph implementation of ELBO-based SVI. The gradient estimator
     is constructed along the lines of reference [1] specialized to the case
     of the ELBO. It supports arbitrary dependency structure for the model
-    and guide as well as baselines for non-reparameteriable random variables.
-    Where possible, dependency information as recorded in the TraceGraph is
-    used to reduce the variance of the gradient estimator.
+    and guide as well as baselines for non-reparameterizable random variables.
+    Where possible, conditional dependency information as recorded in the
+    :class:`Trace` is used to reduce the variance of the gradient estimator.
```
Comment (fritzo) [inline pyro/infer/tracegraph_elbo.py]: ditto:
```
:class:`~pyro.poutine.trace.Trace`
```

File: pyro/infer/tracegraph_elbo.py [hunk 3]
```diff
@@ -170,9 +170,14 @@ class TraceGraph_ELBO(ELBO):
...
+    In particular three kinds of conditional dependency information are
+    used to reduce variance:
+    - the sequential order of samples (z is sampled after y => y does not depend on z)
+    - :class:`~pyro.iarange` contexts
+    - :class:`~pyro.irange` contexts
```
Comment (fritzo) [inline pyro/infer/tracegraph_elbo.py]: nit: 'generators' (rather than 'contexts')?

File: pyro/infer/tracegraph_elbo.py [hunk 4]
```diff
@@ -170,9 +170,14 @@ class TraceGraph_ELBO(ELBO):
...
     of the ELBO. It supports arbitrary dependency structure for the model
-    and guide as well as baselines for non-reparameteriable random variables.
-    Where possible, dependency information as recorded in the TraceGraph is
-    used to reduce the variance of the gradient estimator.
+    and guide as well as baselines for non-reparameterizable random variables.
+    Where possible, conditional dependency information as recorded in the
+    :class:`~pyro.poutine.trace.Trace` is used to reduce the variance of the gradient estimator.
+    In particular three kinds of conditional dependency information are
+    used to reduce variance:
+    - the sequential order of samples (z is sampled after y => y does not depend on z)
```
Comment (jpchen) [inline pyro/infer/tracegraph_elbo.py]: backtick variables

**Pull request merged**